plugins/inputs/hugepages/hugepages.go

```diff
@@ -0,0 +1,292 @@
+	gatherPerNode bool
+	gatherMeminfo bool
+
+	rootHugepagePath string
+	numaNodePath     string
+	meminfoPath      string
+}
+
+func (h *Hugepages) Description() string {
+	return "Gathers huge pages measurements."
+}
+
+func (h *Hugepages) SampleConfig() string {
+	return hugepagesSampleConfig
+}
+
+func (h *Hugepages) Init() error {
+	err := h.parseHugepagesConfig()
+	if err != nil {
+		return err
+	}
+
+	h.rootHugepagePath = rootHugepagePath
+	h.numaNodePath = numaNodePath
+	h.meminfoPath = meminfoPath
+
+	return nil
+}
+
+func (h *Hugepages) Gather(acc telegraf.Accumulator) error {
+	var err error
+
+	if h.gatherRoot {
+		err = h.gatherRootStats(acc)
+		if err != nil {
+			return fmt.Errorf("gathering root stats failed: %v", err)
+		}
+	}
+
+	if h.gatherPerNode {
+		err = h.gatherStatsPerNode(acc)
+		if err != nil {
+			return fmt.Errorf("gathering per node stats failed: %v", err)
+		}
+	}
+
+	if h.gatherMeminfo {
+		err = h.gatherStatsFromMeminfo(acc)
+		if err != nil {
+			return fmt.Errorf("gathering meminfo stats failed: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// gatherStatsPerNode collects root hugepages statistics
+func (h *Hugepages) gatherRootStats(acc telegraf.Accumulator) error {
+	return h.gatherFromHugepagePath(acc, "hugepages_"+rootHugepages, h.rootHugepagePath, hugepagesMetricsRoot, nil)
+}
+
+// gatherStatsPerNode collects hugepages statistics per NUMA node
+func (h *Hugepages) gatherStatsPerNode(acc telegraf.Accumulator) error {
+	nodeDirs, err := ioutil.ReadDir(h.numaNodePath)
+	if err != nil {
+		return err
+	}
+
+	// read metrics from: node*/hugepages/hugepages-*/*
+	for _, nodeDir := range nodeDirs {
+		if !nodeDir.IsDir() || !strings.HasPrefix(nodeDir.Name(), "node") {
+			continue
+		}
+
+		nodeNumber := strings.TrimPrefix(nodeDir.Name(), "node")
+		_, err := strconv.Atoi(nodeNumber)
+		if err != nil {
+			continue
+		}
+
+		perNodeTags := map[string]string{
+			"node": nodeNumber,
+		}
+		hugepagesPath := filepath.Join(h.numaNodePath, nodeDir.Name(), "hugepages")
+		err = h.gatherFromHugepagePath(acc, "hugepages_"+perNodeHugepages, hugepagesPath, hugepagesMetricsPerNUMANode, perNodeTags)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *Hugepages) gatherFromHugepagePath(acc telegraf.Accumulator, measurement, path string, fileFilter map[string]string, defaultTags map[string]string) error {
+	// read metrics from: hugepages/hugepages-*/*
+	hugepagesDirs, err := ioutil.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	for _, hugepagesDir := range hugepagesDirs {
+		if !hugepagesDir.IsDir() || !strings.HasPrefix(hugepagesDir.Name(), "hugepages-") {
+			continue
+		}
+
+		hugepagesSize := strings.TrimPrefix(strings.TrimSuffix(hugepagesDir.Name(), "kB"), "hugepages-")
+		_, err := strconv.Atoi(hugepagesSize)
+		if err != nil {
+			continue
+		}
+
+		metricsPath := filepath.Join(path, hugepagesDir.Name())
+		metricFiles, err := ioutil.ReadDir(metricsPath)
+		if err != nil {
+			return err
+		}
```

Comment:
Do you expect the error messages (especially for the `ReadDir` calls) are sufficiently different to be able to trace errors to the place of occurance? Adding some specific prefixes helps with debugging the code...
```suggestion
		return fmt.Errorf("reading root dir %q failed: %v", path, err)
	}

	for _, hugepagesDir := range hugepagesDirs {
		if !hugepagesDir.IsDir() || !strings.HasPrefix(hugepagesDir.Name(), "hugepages-") {
			continue
		}

		hugepagesSize := strings.TrimPrefix(strings.TrimSuffix(hugepagesDir.Name(), "kB"), "hugepages-")
		_, err := strconv.Atoi(hugepagesSize)
		if err != nil {
			continue
		}

		metricsPath := filepath.Join(path, hugepagesDir.Name())
		metricFiles, err := ioutil.ReadDir(metricsPath)
		if err != nil {
			return fmt.Errorf("reading metric dir %q failed: %v", metricsPath, err)
		}
```

Reply:
I've added "specific prefixes" but without path. Problematic path will be contained in original error.